### PR TITLE
Agent System ID -> UUID

### PR DESCRIPTION
**Pull request by opalmer**
This is the server side change which replaces `Agent.id` with a UUID object and removes `Agent.systemid`.

### CODE REVIEW

File: docs/source/modules/pyfarm.master.api.rst [hunk 1]
```diff
@@ -14,6 +14,7 @@ Submodules
    pyfarm.master.api.pathmaps
    pyfarm.master.api.software
    pyfarm.master.api.tags
+   pyfarm.master.api.tasklogs
```
Comment (guidow) [inline docs/source/modules/pyfarm.master.api.rst]: I don't think these documentation updates should be part of this pull request. They don't even touch on the sysid vs. uuid issue.

File: pyfarm/master/api/agents.py [hunk 2]
```diff
@@ -287,7 +296,7 @@ def post(self):
 
                 if not hasattr(agent, key):
                     return jsonify(
-                        error="Agent has no such column `%s`" % key), \
+                        error="Agent has no such column %r" % key), \
```
Comment (guidow) [inline pyfarm/master/api/agents.py]: This should be in a separate pull request

File: pyfarm/master/api/agents.py [hunk 3]
```diff
@@ -296,7 +305,7 @@ def post(self):
 
                     except Exception as e:
                         return jsonify(
-                            error="Error while setting `%s`: %s" % (key, e)), \
+                            error="Error while setting %r: %s" % (key, e)), \
```
Comment (guidow) [inline pyfarm/master/api/agents.py]: This should be in a separate pull request.

File: pyfarm/master/api/agents.py [hunk 4]
```diff
@@ -547,15 +556,17 @@ def get(self, agent_id):
                 HTTP/1.1 404 NOT FOUND
                 Content-Type: application/json
 
+<<<<<<< HEAD
+                {"error": "Agent `4eefca76-1127-4c17-a3df-c1a7de685541` not "
+                          "found"}
+=======
                 {"error": "Agent 1234 not found"}
+>>>>>>> b0911c70561ee45978af991cc16374e2ce806aea
```
Comment (guidow) [inline pyfarm/master/api/agents.py]: Looks like you had a merge conflict there at some point.

File: pyfarm/master/api/agents.py [hunk 5]
```diff
@@ -620,9 +632,11 @@ def post(self, agent_id):
 
         current_assignments = g.json.pop("current_assignments", None)
         mac_addresses = g.json.pop("mac_addresses", None)
+
         # TODO return BAD_REQUEST on bad mac addresses
         if mac_addresses is not None:
-            mac_addresses = [x.lower() for x in mac_addresses if MAC_RE.match(x)]
+            mac_addresses = [
+                x.lower() for x in mac_addresses if MAC_RE.match(x)]
```
Comment (guidow) [inline pyfarm/master/api/agents.py]: This is a purely cosmetic change that shouldn't be mixed in with a major functionality change.

File: pyfarm/master/api/tasklogs.py [hunk 6]
```diff
@@ -163,7 +163,7 @@ def post(self, job_id, task_id, attempt):
                            error="Specified task not found"), NOT_FOUND
 
         path = join(LOGFILES_DIR, g.json["identifier"])
-        if not realpath(path).startswith(LOGFILES_DIR):
+        if not abspath(path).startswith(LOGFILES_DIR):
```
Comment (guidow) [inline pyfarm/master/api/tasklogs.py]: What's the reason for this change?

(Also, again, this doesn't really look like it belongs in here.)

File: pyfarm/master/api/tasklogs.py [hunk 7]
```diff
@@ -295,7 +295,7 @@ def get(self, job_id, task_id, attempt, log_identifier):
                            error="Specified log not found in task"), NOT_FOUND
 
         path = join(LOGFILES_DIR, log_identifier)
-        if not realpath(path).startswith(LOGFILES_DIR):
+        if not abspath(path).startswith(LOGFILES_DIR):
```
Comment (guidow) [inline pyfarm/master/api/tasklogs.py]: Same comment as for line 166.

File: pyfarm/master/api/tasklogs.py [hunk 8]
```diff
@@ -355,7 +355,7 @@ def put(self, job_id, task_id, attempt, log_identifier):
...
```
Comment (guidow) [inline pyfarm/master/api/tasklogs.py]: Same comment as for line 166.

**Comment by guidow** [comment]
I see you chose to change the id column to a UUID. This is going to be a bit of a challenge to migrate without losing any data... (I know, I was originally the one who suggested that...)

It seems to work after some very superficial tests (starting the master like that and letting the agent announce itself), but I cannot really test this without some database migration work. I'm going to do that on monday.

File: pyfarm/master/api/tasklogs.py (hunk 6 above)
Comment (opalmer) [inline pyfarm/master/api/tasklogs.py]: On a mac, this fixes a bug where `path = join(LOGFILES_DIR, g.json["identifier"])` would not be the start of `realpath(path)` because `/tmp` is a symlink.

You're right however, this should and will be in another change.  And, the better solution is to apply `abspath` to `path` instead I think after looking at this again...

File: pyfarm/master/api/agents.py (hunk 4 above)
Comment (opalmer) [inline pyfarm/master/api/agents.py]: Thanks, should not have missed that one.  Fixed in 0627edb

**Comment by opalmer** [comment]
@guidow, thanks for the review and let me know how the testing goes.  In the mean time I've made the suggested changes and rolled the other unrelated changes into #234 and #325.

**Comment by opalmer** [comment]
Thanks for the update, glad to hear it's working so far.  Sounds like we're ready to merge this though we should probably hold off until we're ready to merge the agent side too.
